pkg/inttest/inttest.go

```diff
@@ -0,0 +1,298 @@
+// Package inttest enables writing of integration tests. Setup Docker containers for dependencies
+// like PostgreSQL, RabbitMQ and AWS S3 (using localstack). Every setup function ensures the
+// container is ready before returning, ensures resources are cleaned up after the tests are
+// finished and return a client ready to interact with the container.
+package inttest
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	amqp "github.com/rabbitmq/amqp091-go"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/dhis2-sre/im-manager/internal/handler"
+	"github.com/dhis2-sre/im-manager/internal/server"
+	"github.com/dhis2-sre/im-manager/pkg/storage"
+	"github.com/gin-gonic/gin"
+	_ "github.com/lib/pq" // postgres driver
+	"github.com/orlangure/gnomock"
+	"github.com/orlangure/gnomock/preset/localstack"
+
+	"github.com/dhis2-sre/im-manager/pkg/config"
+	"github.com/orlangure/gnomock/preset/postgres"
+	"github.com/orlangure/gnomock/preset/rabbitmq"
+
+	"github.com/stretchr/testify/require"
+	"gorm.io/gorm"
+)
+
+// SetupDB creates a PostgreSQL container. Gorm is connected to the DB and runs the migrations.
+func SetupDB(t *testing.T) *gorm.DB {
+	t.Helper()
+
+	container, err := gnomock.Start(
+		postgres.Preset(
+			postgres.WithUser("im", "im"),
+			postgres.WithDatabase("test_im"),
+		),
+	)
+	require.NoError(t, err, "failed to start DB")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop DB") })
+
+	db, err := storage.NewDatabase(config.Postgresql{
+		Host:         container.Host,
+		Port:         container.DefaultPort(),
+		Username:     "im",
+		Password:     "im",
+		DatabaseName: "test_im",
+	})
+	require.NoError(t, err, "failed to setup DB")
+	return db
+}
+
+// SetupS3 creates an S3 container (using localstack) with all the buckets and files in given path.
+func SetupS3(t *testing.T, path string) *S3Client {
+	t.Helper()
+
+	container, err := gnomock.Start(
+		localstack.Preset(
+			localstack.WithServices(localstack.S3),
+			localstack.WithS3Files(path),
+			localstack.WithVersion("2.1.0"),
+		),
+	)
+	require.NoError(t, err, "failed to start S3")
+	t.Cleanup(func() { require.NoError(t, gnomock.Stop(container), "failed to stop S3") })
+
+	return &S3Client{
+		Client: s3.NewFromConfig(
+			aws.Config{
+				Region: "eu-west-1",
```

Comment:
Maybe add a variable for the region, as it's used in two places here. It's not really going to change, though 🤔

Reply:
I am ok with the duplication but extracted a var 👌🏻